Source/FieldSolver/MagnetostaticSolver/MagnetostaticSolver.cpp

```diff
@@ -0,0 +1,462 @@
+    // This routine interpolates to the B-field staggered grid,
+
+    amrex::Array<amrex::MultiFab*, AMREX_SPACEDIM> buf_ptr =
+    {
+#if defined(WARPX_DIM_3D)
+        m_grad_buf_e_stag[lev][0].get(),
+        m_grad_buf_e_stag[lev][1].get(),
+        m_grad_buf_e_stag[lev][2].get()
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+        m_grad_buf_e_stag[lev][0].get(),
+        m_grad_buf_e_stag[lev][2].get()
+#endif
+    };
+
+    // This will grab the gradient values for Ax
+    mlmg[0]->getGradSolution({buf_ptr});
+
+    // Interpolate dAx/dz to By grid buffer, then add to By
+    this->doInterp(m_grad_buf_e_stag[lev][2],
+                   m_grad_buf_b_stag[lev][1]);
+    MultiFab::Add(*(m_b_field[lev][1]), *(m_grad_buf_b_stag[lev][1]), 0, 0, 1, m_b_field[lev][1]->nGrowVect() );
+
+    // Interpolate dAx/dy to Bz grid buffer, then subtract from Bz
+    this->doInterp(m_grad_buf_e_stag[lev][1],
+                   m_grad_buf_b_stag[lev][2]);
+    m_grad_buf_b_stag[lev][2]->mult(-1._rt);
+    MultiFab::Add(*(m_b_field[lev][2]), *(m_grad_buf_b_stag[lev][2]), 0, 0, 1, m_b_field[lev][2]->nGrowVect() );
+
+    // This will grab the gradient values for Ay
+    mlmg[1]->getGradSolution({buf_ptr});
+
+    // Interpolate dAy/dx to Bz grid buffer, then add to Bz
+    this->doInterp(m_grad_buf_e_stag[lev][0],
+                   m_grad_buf_b_stag[lev][2]);
+    MultiFab::Add(*(m_b_field[lev][2]), *(m_grad_buf_b_stag[lev][2]), 0, 0, 1, m_b_field[lev][2]->nGrowVect() );
+
+    // Interpolate dAy/dz to Bx grid buffer, then subtract from Bx
+    this->doInterp(m_grad_buf_e_stag[lev][2],
+                   m_grad_buf_b_stag[lev][0]);
+    m_grad_buf_b_stag[lev][0]->mult(-1._rt);
+    MultiFab::Add(*(m_b_field[lev][0]), *(m_grad_buf_b_stag[lev][0]), 0, 0, 1, m_b_field[lev][0]->nGrowVect() );
+
+    // This will grab the gradient values for Az
+    mlmg[2]->getGradSolution({buf_ptr});
+
+    // Interpolate dAz/dy to Bx grid buffer, then add to Bx
+    this->doInterp(m_grad_buf_e_stag[lev][1],
+                   m_grad_buf_b_stag[lev][0]);
+    MultiFab::Add(*(m_b_field[lev][0]), *(m_grad_buf_b_stag[lev][0]), 0, 0, 1, m_b_field[lev][0]->nGrowVect() );
+
+    // Interpolate dAz/dx to By grid buffer, then subtract from By
+    this->doInterp(m_grad_buf_e_stag[lev][0],
+                   m_grad_buf_b_stag[lev][1]);
+    m_grad_buf_b_stag[lev][1]->mult(-1._rt);
+    MultiFab::Add(*(m_b_field[lev][1]), *(m_grad_buf_b_stag[lev][1]), 0, 0, 1, m_b_field[lev][1]->nGrowVect() );
+
+    // Synchronize B fields
+    for (int idim = 0; idim < 3; idim++)
+    {
+        // Synchronize the ghost cells, do halo exchange
+        ablastr::utils::communication::FillBoundary(*(m_b_field[lev][idim]),
+                                                    m_b_field[lev][idim]->nGrowVect(),
+                                                    WarpX::do_single_precision_comms);
+    }
```

Comment:
This could be removed: the guard cells are refreshed at the beginning of the next PIC iteration anyway ; so I think that this will add unnecessary communications (which comes with a computational cost).
```suggestion
```